man/man8/tcm_nvwrite.8
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "tcm_nvwrite 8"
.TH tcm_nvwrite 8 "2018-07-12" "TCM Management" ""
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
tpm_nvwrite
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBtpm_nvwrite [\s-1OPTIONS\s0]\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBtpm_nvwrite\fR writes data to an \s-1NVRAM\s0 area.
The user must provide the index of the \s-1NVRAM\s0 area as well as the data to
write into the \s-1NVRAM\s0 area. Optionally an offset into the \s-1NVRAM\s0
area can be given where the writing of the data is to start. Similarly
the use may provide the number of bytes to write into the
\&\s-1NVRAM\s0 area. Some \s-1NVRAM\s0 areas only allow write sizes of zero, i.e., \s-1NVRAM\s0
area index 0.
If the \s-1NVRAM\s0 area has permission bits set that require owner or \s-1NVRAM\s0 area
authentication, then the password must be provided via the password option.
.PP
The following options are supported:
.IP "\fB\-h, \-\-help\fR" 4
.IX Item "-h, --help"
Display command usage info.
.IP "\fB\-v, \-\-version\fR" 4
.IX Item "-v, --version"
Display command version info.
.IP "\fB\-l, \-\-log [none|error|info|debug]\fR" 4
.IX Item "-l, --log [none|error|info|debug]"
Set logging level.
.IP "\fB\-u, \-\-unicode\fR" 4
.IX Item "-u, --unicode"
Use \s-1TSM UNICODE\s0 encoding for passwords to comply with applications
using \s-1TSM\s0 popup boxes
.IP "\fB\-z, \-\-well\-known\fR" 4
.IX Item "-z, --well-known"
Use a secret of all zeros (32 bytes of zeros) as either the \s-1NVRAM\s0 area or owner secret.
.IP "\fB\-p, \-\-password\fR" 4
.IX Item "-p, --password"
Depending on what type of authentication is required for the \s-1NVRAM\s0 area, this must either
be the \s-1NVRAM\s0 area or owner password.
.Sp
A password may be directly provided for example by using '\-\-password=password' or
\&\-ppassword. If no password is provided then the program will prompt the user for the
password.
.IP "\fB\-i, \-\-index\fR (optional parameter)" 4
.IX Item "-i, --index (optional parameter)"
The index of the \s-1NVRAM\s0 area.
The parameter may either be a decimal number or a hexadecimal number starting with '0x'.
.IP "\fB\-s, \-\-size\fR" 4
.IX Item "-s, --size"
The number of bytes to write to the \s-1NVRAM\s0 area. This parameter may be used to truncating
the data that are for example read from a file.
The parameter may either be a decimal number or a hexadecimal number starting with '0x'.
.IP "\fB\-n, \-\-offset\fR (optional parameter)" 4
.IX Item "-n, --offset (optional parameter)"
The offset into the \s-1NVRAM\s0 area where to start writing.
The parameter may either be a decimal number or a hexadecimal number starting with '0x'.
.IP "\fB\-f, \-\-filename\fR (optional parameter)" 4
.IX Item "-f, --filename (optional parameter)"
The name of a file from which to read data and copy them into the \s-1NVRAM\s0 area.
.IP "\fB\-d, \-\-data\fR (optional parameter)" 4
.IX Item "-d, --data (optional parameter)"
The data to write into the \s-1NVRAM\s0 area passed in the command line.
.IP "\fB\-m, \-\-fill\-value\fR (optional parameter)" 4
.IX Item "-m, --fill-value (optional parameter)"
The byte to fill the \s-1NVRAM\s0 area with.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBtpm_nvdefine\fR(8), \fBtpm_nvinfo\fR(8), \fBtpm_nvread\fR(8), \fBtpm_nvrelease\fR(8)
.SH "REPORTING BUGS"
.IX Header "REPORTING BUGS"
Report bugs to <trousers\-users@lists.sourceforge.net>


src/tpm_mgmt/tcm_nvwrite.c
/*
 * The Initial Developer of the Original Code is International
 * Business Machines Corporation. Portions created by IBM
 * Corporation are Copyright (C) 2005 International Business
 * Machines Corporation. All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the Common Public License as published by
 * IBM Corporation; either version 1 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * Common Public License for more details.
 *
 * You should have received a copy of the Common Public License
 * along with this program; if not, a copy can be viewed at
 * http://www.opensource.org/licenses/cpl1.0.php.
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <limits.h>

#include "tcm_nvcommon.h"
#include "tcm_tspi.h"
#include "tcm_utils.h"

static BOOL nvindex_set;
static unsigned int nvindex;
static unsigned int offset;
static unsigned int length;
static int fillvalue = -1;
static const char *filename;
static BOOL passWellKnown;
static BOOL askPassword;
static const char *password;
static char *data;

TSM_HCONTEXT hContext = 0;


static int parse(const int aOpt, const char *aArg)
{

	switch (aOpt) {
	case 'i':
		if (parseHexOrDecimal(aArg, &nvindex, 0, UINT_MAX,
				      "NVRAM index") != 0)
			return -1;

		nvindex_set = 1;

		break;

	case 's':
		if (parseHexOrDecimal(aArg, &length, 0, UINT_MAX,
				      "length of data") != 0)
			return -1;
		break;

	case 'n':
		if (parseHexOrDecimal(aArg, &offset, 0, UINT_MAX,
				      "write offset") != 0)
			return -1;
		break;

	case 'd':
		data = strdup(aArg);
		if (data == NULL) {
			logError(_("Out of memory\n"));
			return -1;
		}
		break;

	case 'f':
		filename = aArg;
		break;

	case 'm':
		if (parseHexOrDecimal(aArg, (unsigned int *)&fillvalue,
		                      0, UCHAR_MAX,
				      "fill value") != 0)
			return -1;
		break;

	case 'p':
		password = aArg;
		if (!password)
			askPassword = TRUE;
		else
			askPassword = FALSE;
		passWellKnown =  FALSE;
		break;

	case 'z':
		password = NULL;
		passWellKnown =  TRUE;
		askPassword = FALSE;
		break;

	case 'u':
		useUnicode = TRUE;
		break;

	default:
		return -1;
	}
	return 0;
}

static void help(const char* aCmd)
{
	logCmdHelp(aCmd);
	logUnicodeCmdOption();
	logCmdOption("-z, --well-known",
		     _("Use 32 bytes of zeros (TSM_WELL_KNOWN_SECRET) as the TCM secret authorization data"));
	logCmdOption("-p, --password",
		     _("Owner or NVRAM area password depending on permissions"));
	logNVIndexCmdOption();
	logCmdOption("-s, --size",
		     _("Number of bytes to write to the NVRAM area"));
	logCmdOption("-n, --offset",
		     _("Offset at which to start writing into the NVRAM area"));
	logCmdOption("-f, --filename",
		     _("File whose contents to write into the NVRAM area"));
	logCmdOption("-d, --data",
		     _("Data to write into the NVRAM area"));
	logCmdOption("-m, --fill-value",
		     _("The byte to fill the NVRAM area with"));
}

int main(int argc, char **argv)
{

	TSM_HTCM hTpm;
	TSM_HNVSTORE nvObject;
	TSM_FLAG fNvAttrs;
	UINT32 ulDataLength, bytesToWrite, off;
	BYTE *rgbDataToWrite = NULL;
	TSM_HPOLICY hTpmPolicy, hDataPolicy;
	TCM_NV_DATA_PUBLIC *nvpub = NULL;
	int iRc = -1;
	BYTE well_known_secret[] = TSM_WELL_KNOWN_SECRET;
	int pswd_len = -1;
	struct option hOpts[] = {
		{"index"      , required_argument, NULL, 'i'},
		{"size"       , required_argument, NULL, 's'},
		{"offset"     , required_argument, NULL, 'n'},
		{"data"       , required_argument, NULL, 'd'},
		{"filename"   , required_argument, NULL, 'f'},
		{"fillvalue"  , required_argument, NULL, 'm'},
		{"password"   , optional_argument, NULL, 'p'},
		{"use-unicode",       no_argument, NULL, 'u'},
		{"well-known" ,       no_argument, NULL, 'z'},
		{NULL	 ,       no_argument, NULL, 0},
	};
	struct stat statbuf;
	int fd = -1;
	ssize_t read_bytes;

	initIntlSys();

	if (genericOptHandler
		    (argc, argv, "i:s:n:d:f:m:p::zu", hOpts,
		     sizeof(hOpts) / sizeof(struct option), parse, help) != 0)
		goto out;

	if (nvindex_set == 0) {
		logError(_("You must provide an index for the NVRAM area.\n"));
		goto out;
	}

	if (length > 0 && data == NULL &&
	    filename == NULL &&
	    fillvalue == -1) {
		logError(_("Either data, name of file or fill value must be "
			   "provided.\n"));
		goto out;
	}

	if (data) {
		ulDataLength = strlen(data);

		if (length > 0 && (UINT32)length < ulDataLength)
			ulDataLength = length;

		rgbDataToWrite = (BYTE *)data;
		data = NULL;
	} else if (filename) {
		if (stat(filename, &statbuf) != 0) {
			logError(_("Could not access file '%s'\n"),
				 filename);
			goto out;
		}
		ulDataLength = statbuf.st_size;

		if (length > 0 && (UINT32)length < ulDataLength)
			ulDataLength = length;

		rgbDataToWrite = malloc(ulDataLength);
		if (rgbDataToWrite == NULL) {
			logError(_("Out of memory.\n"));
			return -1;
		}
		fd = open(filename, O_RDONLY);
		if (fd < 0) {
			logError(_("Could not open file %s for reading.\n"));
			return -1;
		}
		read_bytes = read(fd, rgbDataToWrite, ulDataLength);

		if (read_bytes < 0 || ulDataLength != (UINT32)read_bytes) {
			logError(_("Error while reading data.\n"));
			return -1;
		}
		close(fd);
		fd = -1;
	} else if (fillvalue >= 0) {
		if (length < 0) {
			logError(_("Requiring size parameter.\n"));
			return -1;
		}
		ulDataLength = length;
		rgbDataToWrite = malloc(ulDataLength);
		if (rgbDataToWrite == NULL) {
			logError(_("Out of memory.\n"));
			return -1;
		}
		__memset(rgbDataToWrite, fillvalue, ulDataLength);
	} else {
		ulDataLength = 0;
	}

	if (contextCreate(&hContext) != TSM_SUCCESS)
		goto out;

	if (contextConnect(hContext) != TSM_SUCCESS)
		goto out_close;

	if (contextGetTpm(hContext, &hTpm) != TSM_SUCCESS)
		goto out_close;

	fNvAttrs = 0;

	if (contextCreateObject(hContext,
				TSM_OBJECT_TYPE_NV,
				fNvAttrs,
				&nvObject) != TSM_SUCCESS)
		goto out_close;


	if (askPassword) {
		password = _GETPASSWD(_("Enter NVRAM access password: "), &pswd_len,
			FALSE, useUnicode );
		if (!password) {
			logError(_("Failed to get NVRAM access password\n"));
			goto out_close;
		}
	}
	if (password || passWellKnown) {
		if (policyGet(hTpm, &hTpmPolicy) != TSM_SUCCESS)
			goto out_close;

		if (password) {
			if (pswd_len < 0)
				pswd_len = strlen(password);
			if (policySetSecret(hTpmPolicy, strlen(password),
					    (BYTE *)password) != TSM_SUCCESS)
				goto out_close;
		} else {
			if (policySetSecret(hTpmPolicy, TCPA_SM3_256_HASH_LEN,
					    (BYTE *)well_known_secret) != TSM_SUCCESS)
				goto out_close;
		}

		if (contextCreateObject
		    (hContext, TSM_OBJECT_TYPE_POLICY, TSM_POLICY_USAGE,
		     &hDataPolicy) != TSM_SUCCESS)
			goto out_close;

		if (password) {
			if (policySetSecret(hDataPolicy, strlen(password),
					    (BYTE *)password) != TSM_SUCCESS)
				goto out_close;
		} else {
			if (policySetSecret(hDataPolicy, TCPA_SM3_256_HASH_LEN,
					    (BYTE *)well_known_secret) != TSM_SUCCESS)
				goto out_close;
		}

		if (Tspi_Policy_AssignToObject(hDataPolicy, nvObject) !=
		    TSM_SUCCESS)
			goto out_close;
	}

	if (nvindex != TCM_NV_INDEX0) {
		if (getNVDataPublic(hTpm, nvindex, &nvpub)  != TSM_SUCCESS) {
			logError(_("Could not get NVRAM area public information.\n"));
			goto out_close_obj;
		}

		if ((UINT32)offset > nvpub->dataSize) {
			logError(_("The offset is outside the NVRAM area's size of "
			           "%u bytes.\n"),
				 nvpub->dataSize);
			goto out_close_obj;
		}

		if ((UINT32)offset + ulDataLength > nvpub->dataSize) {
			logError(_("Writing of data would go beyond the NVRAM area's size "
			           "of %u bytes.\n"),
				 nvpub->dataSize);
			goto out_close_obj;
		}
	}

	if (Tspi_SetAttribUint32(nvObject,
				 TSM_TSPATTRIB_NV_INDEX,
				 0,
				 nvindex) != TSM_SUCCESS)
		goto out_close_obj;


	bytesToWrite = ulDataLength;
	off = offset;

	if (bytesToWrite == 0 &&
	    NVWriteValue(nvObject, 0, 0, NULL) != TSM_SUCCESS)
		goto out_close_obj;

#define WRITE_CHUNK_SIZE   1024
	while (bytesToWrite > 0) {
		UINT32 chunk = (bytesToWrite > WRITE_CHUNK_SIZE)
			       ? WRITE_CHUNK_SIZE
			       : bytesToWrite;
		if (NVWriteValue(nvObject, off, chunk, &rgbDataToWrite[off-offset])
		    != TSM_SUCCESS)
			goto out_close_obj;

		bytesToWrite -= chunk;
		off += chunk;
	}

	logMsg(_("Successfully wrote %d bytes at offset %d to NVRAM index "
	         "0x%x (%u).\n"),
	       ulDataLength, offset, nvindex, nvindex);

	iRc = 0;

	goto out_close;

      out_close_obj:
	contextCloseObject(hContext, nvObject);

      out_close:
	contextClose(hContext);

      out:
	free(rgbDataToWrite);
	freeNVDataPublic(nvpub);

	return iRc;
}
